### PR TITLE
Mega Man Battle Network 3: Implement New Game

### PR DESCRIPTION
**Pull request by digiholic**
## What is this fixing or adding?
Adds Mega Man Battle Network 3 - Blue for the GameBoy Advance as a playable Archipelago Randomizer

## How was this tested?
Several local solos and async, private multiplayer tests across several seeds. Played at a GDQ side event.
https://www.youtube.com/watch?v=K5db0YyKsow&pp=ygUUZ2RxIG1tYm4zIG11bHRpd29ybGQ%3D

## If this makes graphical changes, please attach screenshots.
image
image
image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,6 +19,7 @@
 *.wixobj
 *.lck
 *.db3
+*.gba
```
Comment (Alchav) [inline .gitignore:22]: I added gba up above while I was adding gb and gbc to the list

**Review by Alchav — CHANGES_REQUESTED** [review]

File: worlds/mmbn3/Rom.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
+from ndspy import lz10
+from Patch import APDeltaPatch
+
+import Utils
+import os
+import hashlib
+import bsdiff4
+
+from .BN3RomUtils import *  # We are literally going to use every single function from this
```
Comment (beauxq) [inline worlds/mmbn3/Rom.py:9]: Even if you use everything from a module, it is still recommended to not use `import *` because it makes it more difficult for linters and type checkers to check **how** you're using them.

File: worlds/mmbn3/docs/setup_en.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Setup Guide for Ocarina of Time Archipelago
```
Comment (beauxq) [inline worlds/mmbn3/docs/setup_en.md:1]:
```suggestion
# Setup Guide for Mega Man Battle Network 3 Archipelago
```

**Comment by ThePhar** [comment]
You'll need to go in and refactor all references to the `MultiWorld` reference from `world` to `multiworld` as that member was renamed in #931 and this will fail on generation once caught up to main otherwise.

**Comment by digiholic** [comment]
Weekend beta testing found an issue with scale. Large numbers of items coming from the network has a tendency to cause the game to crash. Reverting to a draft until the issue is resolved.

**Review by SoldierofOrder — APPROVED** [review]
Only looked at the docs. Great stuff! :)

File: worlds/mmbn3/Rom.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+            raise Exception('Supplied Base Rom does not match US GBA Blue Version.'
+                            'Please provide the correct ROM version')
+
+        get_base_rom_bytes.base_rom_bytes = base_rom_bytes
+    return base_rom_bytes
+
+
+def get_patched_rom_bytes(file_name: str = "") -> bytes:
+    base_rom_bytes = get_base_rom_bytes(file_name)
+    patch_path = os.path.join(os.path.dirname(__file__), "data", "bn3-ap-patch.bsdiff")
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Rom.py:314]: I've got the patch generation code in another repo - https://github.com/digiholic/bn3-ap-patch

If I should include that somewhere in the repository, I can, but I'm not sure where.

File: worlds/mmbn3/docs/en_MegaMan Battle Network 3.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+## What does randomization do to this game?
+
+Items which the player would normally acquire throughout the game have been moved around. Logic remains, so the game is
+always able to be completed, but because of the item shuffle the player may need to access certain areas before they
+would in the vanilla game.
+
+The game begins in "Open Mode", in a story state just before leaving for the WWW Base. All story progress leading up to
+that point have been auto completed. You will be given a starting Style Change upon starting, but will
+have default MegaMan equipped, you can switch to your new style if you want through the Navi menu.
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/docs/en_MegaMan Battle Network 3.md:15]: This one's annoying, the Battle Network series uses `MegaMan`, but every other franchise uses `Mega Man` (except for the times they accidentally use Megaman. Standards are not this series' strong suit) so I have gotten it wrong in a few places. Fixing now.

File: worlds/mmbn3/Options.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+from BaseClasses import MultiWorld
+from Options import Choice, Range, Option, Toggle, DefaultOnToggle, DeathLink, TextChoice
+
+
+class ExtraRanks(Range):
+    """
+    How many extra Undernet Ranks to add to the pool in place of filler items.
+    The more ranks there are, the faster the game will go.
+    Depending on your other options, you might not have enough filler items to replace.
+    If generation errors occur, consider reducing this value.
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Options.py:9]: I've gone back through the items and made a few dozen new filler items. I was playing a bit more literal with the terms "useful" and "filler" so I made some of the less powerful combat options fillers as well.

File: worlds/mmbn3/lz10.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
+import struct
+
+
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/lz10.py:3]: I used ndspy as a reference, but since that was developed for GB, GBA, and DS, not knowing which game it uses, I was able to cut out a lot of the unnecessary checks and alternate cases to use just the GBA algorithm. If it's still too close then I can try to rewrite it, but since it's based off of the algorithm used by the GBA itself it's always going to end up functionally identical.

File: worlds/mmbn3/__init__.py [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+    data_version = 0
+
+    base_id = 0xB31000
+    item_name_to_id: typing.Dict[str, int] = {name: data.code for name, data in item_table.items()}
+    location_name_to_id = {locData.name: locData.id for locData in all_locations}
+
+    web = MMBN3Web()
+
+    def __init__(self, world: MultiWorld, player: int):
+        self.rom_name_available_event = threading.Event()
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/__init__.py:52]: When removed, generation throws an error saying it's not set. So I'm guessing it needs to be there.

File: worlds/mmbn3/BN3RomUtils.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
... 0x79A41C, 0x79A53C, 0x79A628, 0x79A8C0, 0x79AAB4, 0x79AD04, 0x79B134, 0x79B318, 0x79BC18, 0x79BD38, 0x79BE40, 0x79BF58, 0x79C074, 0x79C274, 0x79C400, 0x79C58C, 0x79C818, 0x79C9BC, 0x79CBEC, 0x79D130, 0x79D9CC, 0x79DBEC, 0x79DE34, 0x79E24C, 0x79E39C, 0x79EA5C, 0x79F48C, 0x79FD98, 0x7A0420, 0x7A09A8, 0x7A0B40, 0x7A0DD0, 0x7A128C, 0x7A15BC, 0x7A1910, 0x7A1EE0, 0x7A23C0, 0x7A24A0, 0x7A26C0, 0x7A282C, 0x7A324C, 0x7A356C, 0x7A3C20, 0x7A3D48, 0x7A4424, 0x7A4A28, 0x7A4B70, 0x7A4D30, 0x7A5178, 0x7A5274, 0x7A54D4, 0x7A57F0, 0x7A6250, 0x7A6690, 0x7A68FC, 0x7A7094, 0x7A73E8, 0x7A7FFC, 0x7A85F8, 0x7A8790, 0x7A8EC0, 0x7A900C, 0x7A9908, 0x7AA044, 0x7AA344, 0x7AA7A0, 0x7AA858, 0x7AAC10, 0x7AAE1C, 0x7AAF34, 0x7AB048, 0x7AB2C0, 0x7AB2FC, 0x7AB4A8, 0x7AB5C8, 0x7ABB28, 0x7ABBC8, 0x7ABE48, 0x7ABEA0, 0x7AC32C, 0x7AC478, 0x7AC4F8, 0x7AC5B0, 0x7AC87C, 0x7AC8F8, 0x7ACA4C, 0x7ACD54, 0x7AD3DC, 0x7AD5AC, 0x7ADCF4, 0x7AE20C, 0x7AE3C8, 0x7AEF30, 0x7AF718, 0x7AFBDC, 0x7B2098, 0x7B2864, 0x7B30C8, 0x7B3320, 0x7B34D4, 0x7B3714, 0x7B3834, 0x7B3C74, 0x7B4198, 0x7B4488, 0x7B4A14, 0x7CB3F0, 0x7EA2C8, 0x7EAEB0, 0x7EAFA0, 0x7EB19C, 0x7EB974, 0x7EBAA8, 0x7EC600, 0x7EF468, 0x7F0B40, 0x7F95F8, 0x7F961C, 0x7F98F0, 0x7FAC00]
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/BN3RomUtils.py:6]: It's future proofing in case I need to change the text of any of them. Currently, there are no items in any uncompressed archives, but I believe there are some shops in these address spaces which I hope to add functionality to later. I can comment this out for now, but it'll be hand to have later so it can check by address if an archive should be compressed or not.

File: worlds/mmbn3/Items.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+    count: typing.Optional[int] = 1
+    recipient: typing.Optional[str] = "Myself"
+
+
+class MMBN3Item(Item):
+    game: str = "MegaMan Battle Network 3"
+    
+
+keyItemList: typing.List[ItemData] = [
+    ItemData(0xB31000, ItemName.Progressive_Undernet_Rank, ItemClassification.progression, ItemType.Undernet, 27),
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Items.py:53]: Did not know what that option meant but you're right, it does.

File: worlds/mmbn3/Items.py [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+class ItemData(typing.NamedTuple):
+    code: int
+    itemName: str
+    progression: ItemClassification
+    type: str
+    itemID: typing.Optional[int] = 0x00
+    subItemID: typing.Optional[int] = 0x00
+    count: typing.Optional[int] = 1
+    recipient: typing.Optional[str] = "Myself"
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Items.py:45]: Holdover from before I had data to populate all these. Fixed.

File: worlds/mmbn3/Items.py [hunk 12]
```diff
@@ -0,0 +1,351 @@
+import typing
+
+from BaseClasses import Item, ItemClassification
+from .Names import ItemName
+
+
+class ItemType:
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Items.py:8]: Dangit, missed one. Fixed.

File: worlds/mmbn3/Locations.py [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+
+
+def setup_locations(world, player: int):
+    # If we later include options to change what gets added to the random pool,
+    # this is where they would be changed
+    location_table = {locData.name: locData.id for locData in all_locations}
+    return location_table
+
+
+lookup_id_to_name: typing.Dict[int, str] = {locData.id: locData.name for locData in all_locations}
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/Locations.py:358]: `setup_locations` is another one of those things that bricks generation if it's removed. The other two are actually unused though, so they've been removed.

File: worlds/mmbn3/lz10.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/lz10.py:8]: Rewritten using an MIT compatible LZ10 library, https://github.com/magical/nlzss

Since it only operates on files instead of byte streams, I had to do a bit of tweaking from the source code in order to get it to work with this use case, but I got the benefit of cutting out all the LZS11 stuff as well so it's smaller than just including that library would have been in the first place.

File: worlds/mmbn3/__init__.py [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+        for itemName in required_items:
+            self.multiworld.itempool.append(self.create_item(itemName))
+
+        # Then, get a random amount of fillers until we have as many items as we have locations
+        filler_items = []
+        for item in all_items:
+            if item.progression == ItemClassification.filler:
+                freq = player_item_frequencies[self.player][item.itemName] \
+                    if self.player in player_item_frequencies and item.itemName in player_item_frequencies[self.player] \
+                    else item_frequencies[item.itemName] if item.itemName in item_frequencies else 1
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/__init__.py]: The item frequencies for the filler items is the same as the others, it's how many items there are in the game, but some of them won't end up fitting since their locations have been replaced with things like the progressive undernet ranks that normally come from story progress instead of item checks. So it's more like it's randomly deciding which filler items to not include in the generation. So, if a filler item has a frequency of 4, that means there are four of them in the pool that all _could_ get in, but might not. I see now that I never actually remove them from the list if they're chosen, so I'm adding in that functionality now, it should be have more like a normal item now.

File: worlds/mmbn3/BN3RomUtils.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+
+
+def generate_text_bytes(message) -> bytearray:
+    byte_list = []
+    for c in message:
+        byte_list.append(charDict[c])
+    return bytearray(byte_list)
+
+
+def generate_chip_get(chip, code, amt) -> bytearray:
```
[the remark this answers is not in the document]
Comment (digiholic) [reply worlds/mmbn3/BN3RomUtils.py:61]: Just to make the function match the one in the game's bytecode, which takes an amount even though almost every event that gives chips, programs, or key items always uses 1. I could replace with a constant if necessary, but is it actually necessary?

File: host.yaml [hunk 16]
```diff
@@ -154,4 +154,7 @@ smw_options:
   # Set this to false to never autostart a rom (such as after patching)
   # True for operating system default program
   # Alternatively, a path to a program to open the .sfc file with
-  rom_start: true
```
Comment (00Bren) [inline host.yaml]: Was this an intentional removal? Seems like this could impact smw?

File: worlds/mmbn3/Names/LocationName.py [hunk 17]
```diff
@@ -0,0 +1,275 @@
+## Blue Mystery Datas
+# ACDC Area
+ACDC_1_Southwest_BMD = "ACDC 1 Southwest BMD"
```
Comment (00Bren) [inline worlds/mmbn3/Names/LocationName.py]: Nit: Instead of writing these out like this, adding them to a dictionary (or maybe a few different dictionaries) would take up a similar amount of space in this file, and allow you to enumerate over them much faster in other files (instead of having to write them out in multiple places).

File: host.yaml (hunk 16 above)
Comment (digiholic) [reply host.yaml]: I didn't make this change, it came from merging in changes from main. Checking the main repository, it's missing there as well, so I'm guessing it's okay?

File: worlds/mmbn3/Names/LocationName.py (hunk 17 above)
Comment (digiholic) [reply worlds/mmbn3/Names/LocationName.py]: The idea of this is to have the names be compile-time constants to avoid typos causing issues. I'm not sure how a dictionary would solve that issue, as I'd still need compile-time constants for each key in the dictionary mapping to a location name.

Comment (00Bren) [reply worlds/mmbn3/Names/LocationName.py]: containing in a dictionary/map just allows you to iterate over some of the items (which is done in a few spots like Regions.py), while also retaining the ability to search for individual variable names. For example you could do something like : 
###PseudoCode

all acdc regions = {
regionVar1 = "regionName1", 
regionVar2 = "regionName2"
etc
}

all scilab regions = {...}
etc

allRegions.addAll(acdcRegions)
allRegions.addAll(scilabRegions)

Then you can iterate over the scilab areas for certain things that need iteration (like Regions.py), and then still access vars in allRegions.

But, as I prefaced my original comment with, this is just me being nitpicky more than anything (I think it'd cut down ~250-300 lines of code).

File: host.yaml (hunk 16 above)
Comment (00Bren) [reply host.yaml]: Oh, weird that it's missing there and still shows that this is a removal. Yeah, if it's missing there I would imagine this isn't an issue, but I'm not sure.

File: worlds/LauncherComponents.py [hunk 18]
```diff
@@ -97,6 +97,9 @@ def __call__(self, path: str):
               file_identifier=SuffixIdentifier('.apzl')),
     # Kingdom Hearts 2
     Component('KH2 Client', "KH2Client"),
+
+    #MegaMan Battle Network 3
+    Component('MMBN3 Client', 'MMBN3Client', file_identifier=SuffixIdentifier('.abpn3'))
```
Comment (Zunawe) [inline worlds/LauncherComponents.py:102]: abpn3

File: worlds/mmbn3/docs/setup_en.md [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+Double-click on your `.apbn3` file to start your client and start the ROM patch process. Once the process is finished
+(this can take a while), the client and the emulator will be started automatically (if you associated the extension
+to the emulator as recommended).
+
+### Connect to the Multiserver
+
+Once both the client and the emulator are started, you must connect them. Within the emulator click on the "Tools"
+menu and select "Lua Console". Click the folder button or press Ctrl+O to open a Lua script.
+
+Navigate to your Archipelago install folder and open `data/lua/mmbn3/mmbn3_connector.lua`.
```
Comment (Zunawe) [inline worlds/mmbn3/docs/setup_en.md:67]: This path is out of date after the lua changes

File: MMBN3Client.py [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+        "debug": debugEnabled
+        })
+
+
+async def parse_payload(payload: dict, ctx: MMBN3Context, force: bool):
+    # Game completion handling
+    if payload["gameComplete"] and not ctx.finished_game:
+        await ctx.send_msgs([{
+            "cmd": "StatusUpdate",
+            "status": 30
```
Comment (Zunawe) [inline MMBN3Client.py:158]: You can use `ClientStatus.CLIENT_GOAL` from `NetUtils.py` instead of `30` here. Would be good for readability and future-proofing

File: data/lua/mmbn3_connector.lua [hunk 21]
```diff
@@ -0,0 +1,1065 @@
...
+        saveItemIndexToRAM(0)
+    end
+    return last_index
+end
+
+local loadPlayerNameFromROM = function()
+    return memory.read_bytes_as_array(0x7FFFC0,63,"ROM")
+end
+
+-- TODO pass the whole chunk of data and check locations on the other end
```
Comment (Zunawe) [inline data/lua/mmbn3_connector.lua:142]: Would highly recommend doing this if it's not too much trouble (I don't know exactly how these flags are laid out, but they look fairly compact). Lots of duplicated information and multiple reads of the same byte below. And if this is ever phased out in favor of SNI, you'll already have that code in python.

File: worlds/mmbn3/__init__.py [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+
+    def generate_basic(self) -> None:
+        """
+        called after the previous steps. Some placement and player specific randomizations can be done here. After this
+        step all regions and items have to be in the MultiWorld's regions and itempool.
+        """
+        # place "Victory" at "Final Boss" and set collection as win condition
+        self.multiworld.get_location(LocationName.Alpha_Defeated, self.player) \
+            .place_locked_item(self.create_event(ItemName.Victory))
+        self.multiworld.completion_condition[self.player] = \
```
Comment (Jarno458) [inline worlds/mmbn3/__init__.py:376]: setting completion rule should be done in set_rules()

**Review by Jarno458 — CHANGES_REQUESTED** [review]

**Review by Jarno458 — APPROVED** [review]

File: data/lua/mmbn3_connector.lua (hunk 21 above)
Comment (Zunawe) [reply data/lua/mmbn3_connector.lua]: Awesome, looks much nicer.

**Review by ScipioWright — APPROVED** [review]
Only real comment is the little tutorial comments in generate_early and create_regions probably can get removed, since those look like they're more for the worlds doc template things rather than being in the full world.

File: worlds/mmbn3/__init__.py [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+                        state.has(ItemName.Progressive_Undernet_Rank, self.player, 8)
+                region.exits.append(entrance)
+
+    def create_items(self) -> None:
+        # First add in all progression and useful items
+        required_items = []
+        for item in all_items:
+            if item.progression != ItemClassification.filler:
+                freq = self.item_frequencies.get(item.itemName, 1)
+                required_items += [item.itemName] * freq
```
Comment (ThePhar) [inline worlds/mmbn3/__init__.py:138]: Do not use this kind of syntax for creating multiple items. This ends up duplicating the same reference rather than creating a new instance. It might not be terrible in your use case, but I'd recommend this instead just to be in the habit:

```suggestion
                required_items += [item.itemName for _ in range(freq)]
```

File: worlds/mmbn3/__init__.py [hunk 24]
```diff
@@ -0,0 +1,486 @@
...
+
+        for itemName in required_items:
+            self.multiworld.itempool.append(self.create_item(itemName))
+
+        # Then, get a random amount of fillers until we have as many items as we have locations
+        filler_items = []
+        for item in all_items:
+            if item.progression == ItemClassification.filler:
+                freq = self.item_frequencies.get(item.itemName, 1)
+                filler_items += [item.itemName] * freq
```
Comment (ThePhar) [inline worlds/mmbn3/__init__.py:148]: Ditto as reason above.

```suggestion
                filler_items += [item.itemName for _ in range(freq)]
```

File: worlds/mmbn3/BN3RomUtils.py [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+        return item_name
+    # Next, PascalCase it
+    item_name = item_name.replace("_", " ").replace("-", " ").title().replace(" ", "")
+    if len(item_name) <= maxLength:
+        return item_name
+    # If it's still too long, start removing vowels till it's short enough or we run out
+    while len(item_name) > maxLength and any(c in item_name for c in ['a', 'e', 'i', 'o', 'u']):
+        item_name = re.sub("[aeiou]", "", item_name, 1)
+    # If it's somehow still too long, truncate and return
+    return item_name[0:maxLength]
```
Comment (black-sliver) [inline worlds/mmbn3/BN3RomUtils.py:194]: missing newline at the end of the file and some (most?) of the following files.

File: Launcher.py [hunk 26]
```diff
@@ -82,7 +82,6 @@ def open_folder(folder_path):
     else:
         webbrowser.open(folder_path)
 
-
```
Comment (black-sliver) [inline Launcher.py:85]: why is this here?

**Review by ThePhar — CHANGES_REQUESTED** [review]
In addition to the below, I did notice your generate output takes my computer about 8 seconds per player to output. While it might be fine, I'm hesitant to approve the PR unless that time can be improved.

**Review by ThePhar — APPROVED** [review]
After talking with @Berserker66, we should be fine for merging as is. If possible though @digiholic, if we can try to improve the (de)compression algo before the release after next that would be ideal.

**Comment by digiholic** [comment]
Yes, I'm looking into doing the compression externally with a faster library. Barring that, I'll see if I can multithread it or try to find out what specifically is making nlzss so much slower than ndspy's
